### PR TITLE
Add callable signatures to type annotations.

### PR DESCRIPTION
**Pull request by erniejunior**
## Description
I meant to review TD3 but ended up adding the signature to all callable type annotations in the codebase.

## Motivation and Context
Adding the signature to callables in the type annotations improves documentation and gives a much better idea what kind of function must be passed.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [ ] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

**Comment by erniejunior** [comment]
Note: I could not run `make pytest` locally but just executing pytest with my IDE worked just fine.

File: stable_baselines3/a2c/a2c.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def __init__(
         self,
         policy: Union[str, Type[ActorCriticPolicy]],
         env: Union[GymEnv, str],
-        learning_rate: Union[float, Callable] = 7e-4,
+        learning_rate: Union[float, Callable[[float], float]] = 7e-4,
```
Comment (araffin) [inline stable_baselines3/a2c/a2c.py:58]: Maybe add that to the type aliases? (as it comes quite often)
for the name, I'm not sure, maybe `LearningRateOrSchedule`

File: stable_baselines3/common/utils.py [hunk 2]
```diff
@@ -88,7 +88,7 @@ def get_schedule_fn(value_schedule: Union[Callable, float]) -> Callable:
     return value_schedule
 
 
-def get_linear_fn(start: float, end: float, end_fraction: float) -> Callable:
+def get_linear_fn(start: float, end: float, end_fraction: float) -> Callable[[float], float]:
```
Comment (araffin) [inline stable_baselines3/common/utils.py:91]: same, maybe define a `Schedule` type alias

**Review by araffin — COMMENTED** [review]
Just small remarks about new type aliases, other than that LGTM ;)

File: stable_baselines3/common/utils.py (hunk 2 above)
Comment (erniejunior) [reply stable_baselines3/common/utils.py:91]: I will go with `LearningRateSchedule` since we might have other kinds of schedules in the future.

Comment (araffin) [reply stable_baselines3/common/utils.py:91]: in fact, for PPO, there is already the clipping ratio which could be a schedule (currently not implemented)
and in DQN, you have the exploration schedule for the epsilon-greedy policy

**Comment by erniejunior** [comment]
What do you think about documenting the type aliases? This would be a nice place to state how the learning rate schedule work (inputs ranging from 0 to 1 etc). I could add it and also add the documentation from the `LearningRateSchedule` but I do not feel confident about documenting the other type aliases right now.

**Comment by araffin** [comment]
>What do you think about documenting the type aliases?

Where do you want to put that? Comments in the file? or a file under `docs/` folder?
Yes, more documentation is always good ;)

File: stable_baselines3/sac/policies.py [hunk 3]
```diff
@@ -16,6 +16,8 @@
 )
 
 # CAP the standard deviation of the actor
+from stable_baselines3.common.type_aliases import LearningRateSchedule
```
Comment (araffin) [inline stable_baselines3/sac/policies.py:19]: why is it not at the top?

File: stable_baselines3/td3/td3.py [hunk 4]
```diff
@@ -21,7 +21,7 @@ class TD3(OffPolicyAlgorithm):
     Paper: https://arxiv.org/abs/1802.09477
     Introduction to TD3: https://spinningup.openai.com/en/latest/algorithms/td3.html
 
-    :param policy: The policy model to use (MlpPolicy, CnnPolicy, ...)
+    :param policy: The policy model to use (MlpPolicy, CnnPolicy, ...)  # TODO(max): is this true, the type annotations say otherwise
```
Comment (araffin) [inline stable_baselines3/td3/td3.py:24]: `MlpPolicy` is an alias of `TD3Policy` and `CnnPolicy` derives from `TD3Policy`

**Comment by erniejunior** [comment]
I wanted to add a new type aliases section in common that just uses `..automodule::` to document the `type_aliases.py`. However I could not get it to work due to the following problem:

This is the docstring I use in `type_aliases.py`
```python
LearningRateSchedule = Callable[[float], float]
"""This is the learning rate schedule"""
```
However, this will only render `alias of typing.Callable` in HTML and not `This is the learning rate schedule` as expected. I only get it to render properly when I change the type alias to:
```python
LearningRateSchedule = Callable
```
without the signature. Do you have any idea for why this could be? I think I saw some issues where you were using some tricks to get the type annotations properly rendered in sphinx?

**Comment by araffin** [comment]
>However, this will only render alias of typing.Callable in HTML and not This is the learning rate schedule as expected. I only get it to render properly when I change the type alias to:

I'm afraid there is no way around... For instance, I couldn't get it to work for the policies.
At least, it is documenting in the developer guide.

**Review by araffin — APPROVED** [review]
LGTM, thanks =)
